src/main/java/org/opensearch/performanceanalyzer/rca/store/rca/hotshard/HotShardRca.java

```diff
@@ -106,108 +96,151 @@ private void consumeFlowUnit(
+                        usageWindow = new SummarizedWindow();
+                        summarizationMap.put(indexShardKey, usageWindow);
                     }
-                    usageDeque.next(new SlidingWindowData(this.clock.millis(), usage));
+                    usageWindow.next(metricType, usage, this.clock.millis());
                 }
             } catch (Exception e) {
                 PerformanceAnalyzerApp.RCA_VERTICES_METRICS_AGGREGATOR.updateStat(
                         RcaVerticesMetrics.HOT_SHARD_RCA_ERROR, "", 1);
-                LOG.error("Failed to parse metric in FlowUnit: {} from {}", record, metricType, e);
+                LOG.error(
+                        "Failed to parse metric in FlowUnit: {} from {}",
+                        record,
+                        metricType.toString(),
+                        e);
             }
         }
     }
 
-    private void consumeMetrics(
-            final Metric metric,
-            final HashMap<IndexShardKey, SlidingWindow<SlidingWindowData>> metricMap) {
+    private void consumeMetric(Metric metric, AllMetrics.OSMetrics metricType) {
         for (MetricFlowUnit metricFlowUnit : metric.getFlowUnits()) {
             if (metricFlowUnit.getData() != null) {
-                consumeFlowUnit(metricFlowUnit, metric.getClass().getName(), metricMap);
+                consumeFlowUnit(metricFlowUnit, metricType);
             }
         }
     }
 
-    private double fetchUsageValueFromMap(
-            HashMap<IndexShardKey, SlidingWindow<SlidingWindowData>> usageMap,
-            IndexShardKey indexShardKey) {
-        double value = 0;
-        if (usageMap.get(indexShardKey) != null) {
-            value = usageMap.get(indexShardKey).readAvg(TimeUnit.SECONDS);
+    private void checkAgainstThreshold(
+            MinMaxPriorityQueue<NamedSummarizedWindow> queue,
+            Map.Entry<IndexShardKey, SummarizedWindow> entry,
+            AllMetrics.OSMetrics metricType,
+            double threshold) {
+        // TODO : Check if null can break out here
+        double metricValue = entry.getValue().readAvgMetricValue(TimeUnit.SECONDS, metricType);
+        if (metricValue > threshold) {
+            queue.add(new NamedSummarizedWindow(entry.getValue(), entry.getKey()));
+
+            LOG.debug(
+                    "Hot Shard Identified, Shard : {} , metricValue = {} , metricThreshold = {}, metricType = {}",
+                    entry.getKey().toString(),
+                    metricValue,
+                    threshold,
+                    metricType.toString());
+        }
+    }
+
+    private void drainQueue(
+            MinMaxPriorityQueue<NamedSummarizedWindow> queue,
+            Map<IndexShardKey, SummarizedWindow> consumersToSend) {
+        while (!queue.isEmpty()) {
+            NamedSummarizedWindow candidate = queue.remove();
+            if (consumersToSend.containsKey(candidate.indexShardKey)) {
+                continue;
```

Comment:
Actually if the key is already contained, it will point to the same value (SummarizedWindow) we're looking at here, so the information is already preserved, this way we also recycle references to the Windows and have no new structure creations so it's a win-win.

Reply:
![autodraw 2_23_2023(1)](https://user-images.githubusercontent.com/86674853/220902103-a56fca75-5407-47be-8a2f-d3bfb661dfd5.png)

Reply:
Thank you for explaining this. 

As a follow-up, we should do a README for Hot Shards RCA and we can add this image and the details there for users to understand the implementation better.